modules/contact/include/userobjects/PenaltyFrictionUserObject.h

```diff
@@ -0,0 +1,84 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "WeightedVelocitiesUserObject.h"
+
+template <typename>
+class MooseVariableFE;
+
+/**
+ * User object that computes tangential pressures due to friction using a penalty approach
+ */
+class PenaltyFrictionUserObject : public WeightedVelocitiesUserObject
+{
+public:
+  static InputParameters validParams();
+
+  PenaltyFrictionUserObject(const InputParameters & parameters);
+
+  virtual const ADVariableValue & contactPressure() const override;
+  virtual const ADVariableValue & contactTangentialPressureDirOne() const override;
+  virtual const ADVariableValue & contactTangentialPressureDirTwo() const override;
+
+  virtual void initialize() override;
+  virtual void finalize() override;
+  virtual void reinit() override;
+  virtual void timestepSetup() override;
+
+  Real getNormalContactPressure(const Node * const node) const;
+  Real getFrictionalContactPressure(const Node * const node, const unsigned int component) const;
+  Real getAccumulatedSlip(const Node * const node, const unsigned int component) const;
+  Real getTangentialVelocity(const Node * const node, const unsigned int component) const;
+  Real getNormalWeightedGap(const Node * const node) const;
+
+protected:
+  virtual const VariableTestValue & test() const override;
+  virtual bool constrainedByOwner() const override { return false; }
+
+  /// The normal penalty factor
+  const Real _penalty;
```

Comment:
It really sucks that we have this pattern where we always have to duplicate code between XXXFrictionUserObject and XXXWeightedGapUserObject. Would it make sense to have `WeightedVelocitiesUserObject` inherit from `T` as opposed to `WeightedGapUserObject`?